### PR TITLE
Add graph walking functions to Graph and GraphOperation

### PR DESCRIPTION
**Pull request by rnett**
Extracted from #211, adds graph walking and a minimal subgraph function.

Also fixes a bug in Constant where the scope wasn't applied (including device).

### CODE REVIEW

**Comment by rnett** [comment]
I may need to add a flag to include variables, too.

**Comment by rnett** [comment]
The CI seems to be broken again, I'm not having issues running `install` for the whole project locally, but there's javadoc resolution errors from framework.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java [hunk 1]
```diff
@@ -138,14 +217,101 @@ public GraphOperation operation(String name) {
...
+   * @param outputs the outputs of the subgraph
+   * @param allowedBodyOps types of ops to allow as 0-input ops in the body.  Allows all (except {@code
+   * forbiddenBodyOps}) if null.
+   * @param forbiddenBodyOps types of ops to never allow as 0-input ops in the body.  Forbids all (except {@code
+   * allowedBodyOps}) if null.
+   * @return the set of operations needed to calculate outputs from inputs, including outputs and inputs
+   * @throws IllegalStateException if outputs depends on ops outside of the subgraph (i.e. is not calculable based
+   * solely on inputs)
+   */
+  public synchronized Set<GraphOperation> completeSubgraph(Set<Operand<?>> inputs, Set<Operand<?>> outputs,
```
Comment (Craigacp) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:241]: Should we have an overload for this that just accepts the first two arguments and calls this with `Collections.emptySet()` for those arguments? Or document that all the arguments must be non-null and expect our users to suppled `Collections.emptySet()` on their own?

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java [hunk 2]
```diff
@@ -138,14 +217,101 @@ public GraphOperation operation(String name) {
...
+    if (forbiddenBodyOps != null && allowedBodyOps != null) {
+      for (String t : forbiddenBodyOps) {
+        if (allowedBodyOps.contains(t)) {
+          throw new IllegalArgumentException("Can't allow and forbid op type " + t + ".");
+        }
+      }
+    }
+
+    Queue<GraphOperation> currents = new LinkedList<>();
+    Set<GraphOperation> seen = new LinkedHashSet<>(outputs.size());
```
Comment (Craigacp) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:253]: `seen` is allocated as `outputs.size()`, but we immediately add each element of `inputs` to it. Should it be sized as `inputs.size()` (or even `inputs.size() + outputs.size()`)?

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java [hunk 3]
```diff
@@ -138,14 +217,101 @@ public GraphOperation operation(String name) {
...
+    while (!currents.isEmpty()) {
+      GraphOperation op = currents.poll();
+
+      // skip if already present
+      if (!seen.add(op)) {
+        continue;
+      }
+
+      if (op.numControlInputs() + op.numInputs() == 0) {
+        if ((forbiddenBodyOps != null && forbiddenBodyOps.contains(op.type()))
```
Comment (Craigacp) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:279]: Initialising these to `Collections.emptySet()` internally even if we don't make users do it will make this if a lot less ugly.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphOperation.java [hunk 4]
```diff
@@ -170,6 +184,121 @@ Tensor tensor(int outputIdx) {
...
+
+      TF_OperationAllInputs(getUnsafeNativeHandle(), handles, numInputs);
+
+      List<Operand<?>> operands = new ArrayList<>(numInputs);
+      for (int i = 0; i < numInputs; i++) {
+        TF_Output atPos = handles.position(i);
+        TF_Operation op = atPos.oper();
+        int index = atPos.index();
+        String opName = TF_OperationName(op).getString();
+        operands.add(graph.operationOrError(opName).output(index));
```
Comment (Craigacp) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphOperation.java:207]: In what situations could this throw?

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphOperation.java [hunk 5]
```diff
@@ -170,6 +184,121 @@ Tensor tensor(int outputIdx) {
     throw new IllegalStateException("Graph tensors must be fetched by running a session");
   }
 
+
+  public int numInputs() {
```
Comment (Craigacp) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphOperation.java:188]: Javadoc for these new methods please.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphOperation.java [hunk 6]
```diff
@@ -170,6 +184,121 @@ Tensor tensor(int outputIdx) {
...
+      TF_Input handles = new TF_Input(numConsumers);
+
+      TF_OperationOutputConsumers(output, handles, numConsumers);
+
+      Set<GraphOperation> operands = new LinkedHashSet<>(numConsumers);
+      for (int i = 0; i < numConsumers; i++) {
+        TF_Input atPos = handles.position(i);
+        TF_Operation op = atPos.oper();
+        String opName = TF_OperationName(op).getString();
+        operands.add(graph.operationOrError(opName));
```
Comment (Craigacp) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphOperation.java:233]: When could this one throw?

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/Constant.java [hunk 7]
```diff
@@ -1355,13 +1356,15 @@
...
   @Endpoint(name = "constantOf")
   public static <T extends TType> Constant<T> create(Scope scope, T tensor) {
-    return new Constant<>(
-        scope
-            .env()
-            .opBuilder("Const", scope.makeOpName("Const"))
-            .setAttr("value", tensor)
-            .setAttr("dtype", tensor.dataType())
-            .build());
+    OperationBuilder builder = scope
```
Comment (Craigacp) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/Constant.java:1359]: Why is this changed?

**Review by Craigacp — COMMENTED** [review]
There are a lot of unnecessary formatting changes in this PR. Could you back those out?

**Comment by Craigacp** [comment]
It might be useful to have two extra graph walking functions one which traces all the ops which depend on a set of inputs, and one which traces all the ops which are depended on by a set of outputs. That way you can take any graph and figure out what you need to feed into it for a given output. That's not likely to be in a production system, but it would be super useful when poking a model in jshell. I've had to write ad-hoc things which do that when I was trying to understand the TF 1.x graph structure for saving, and having a supported method would be nice.

**Comment by rnett** [comment]
> It might be useful to have two extra graph walking functions one which traces all the ops which depend on a set of inputs, and one which traces all the ops which are depended on by a set of outputs. That way you can take any graph and figure out what you need to feed into it for a given output. That's not likely to be in a production system, but it would be super useful when poking a model in jshell. I've had to write ad-hoc things which do that when I was trying to understand the TF 1.x graph structure for saving, and having a supported method would be nice.

Yeah, I can do that.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/Constant.java (hunk 7 above)
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/Constant.java:1359]: Because it wasn't applying the scope correctly (missed the `scope.apply` call), control dependencies and the device were missed.  The control deps are what caused the issue here, but I ran into the device elsewhere as well.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java [hunk 8]
```diff
@@ -128,6 +136,77 @@ public GraphOperation operation(String name) {
...
 
+  /**
+   * Returns the operation (node in the Graph) with the provided name, or throws {@link IllegalArgumentException} if
+   * there isn't one.
+   *
+   * @param name name of the operation to look for
+   * @return operation in the graph with this name
+   * @throws IllegalArgumentException if no such operation exists in the Graph
+   */
+  public GraphOperation operationOrError(String name) {
```
[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:147]: Cause that's a lot of boilerplate every time you want to get a for-sure existing op.  There's multiple places where it's used currently (graph walking and session, mostly), so we'd have utility functions for it anyways.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java (hunk 1 above)
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:241]: The reason I made it take nulls in the first place is because calling with both as emptySet errors whenever something is in the body (since we're allowing the empty set of ops).  I think an overload is a good idea though, and specifying the first two args must be non-null.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java (hunk 3 above)
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:279]: See one of the other comments, nulls are necessary to distinguish from "only allow the empty set" which would error whenever something is in the body.  Optionally we could do that behavior (i.e. allow everything) for the empty set, but imo nulls is clearer (and what if that's what you want?).

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:147]: A lot of the uses cases are places where we require the op to exist, and it should always exist, but if it somehow doesn't throwing the IllegalStateException is clearer and easier to debug than an NPE.

**Comment by rnett** [comment]
> There are a lot of unnecessary formatting changes in this PR. Could you back those out?

This is due to my auto format on commit, so not really, at least not without it being a big pain.  It's google-java-format, though, so it should really have already been done.  Imo we should add a format check for PRs in the CI (and format everything existing to match), at least once it's not having time issues.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java (hunk 3 above)
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:279]: Why would we want the behaviour of only allowing the empty set? So you could catch the runtime exception and have different behaviour if there were intermediate nodes?

Anything that implies users have to catch a subclass of `RuntimeException` isn't good.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:147]: As for the name, imo `OrError` sounded/read better than `OrException`, although now that I think about it `OrThrow` seems fine.  Or we could make these the default methods and use `OrNull` for the nullable ones.  Thoughts?

[the remark this answers is not in the document]
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:147]: I think I prefer `orThrow`, as that's more similar to the methods in the JDK - https://docs.oracle.com/en/java/javase/15/docs/api/java.base/java/util/Optional.html#orElseThrow(). It's worth noting that if we know the exception can't happen, then the one that returns null will likely be better optimised by the JIT compiler because exceptions get in the way.

[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:147]: That makes sense.  Currently the only uses of the throwing variants are with user input, or in the graph walking functions.  The only case where the graph walking functions would have issues is if the graph is changed in another thread and the inputs were removed (which we don't even expose APIs for) so I'll go ahead and convert them to the non-throwing variant.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java (hunk 3 above)
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:279]: It's for 0 input intermediate nodes specifically, i.e. things that don't depend on the inputs.  Generally `Constant` is ok and the others (`Placeholder`, `Variable`, etc) are ok, but since I'm making it a parameter "none" is an option.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java [hunk 9]
```diff
@@ -224,33 +226,62 @@ private GraphOperation graphOp(Operand<?> operand) {
...
 
+    if (inputs == null) {
+      throw new IllegalArgumentException("Inputs can't be null.");
+    }
+
+    if (outputs == null) {
+      throw new IllegalArgumentException("Outputs can't be null.");
+    }
+
     Queue<GraphOperation> currents = new LinkedList<>();
```
Comment (Craigacp) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:283]: Given this is an `ArrayDeque` in the other new methods, maybe this one should be an `ArrayDeque` not a `LinkedList`

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java [hunk 10]
```diff
@@ -304,6 +336,80 @@ private GraphOperation graphOp(Operand<?> operand) {
...
   }
 
+  /**
+   * Get all ops directly or indirectly required to calculate {@code outputs} (not including {@code outputs}), including
+   * control dependencies.
+   *
+   * @param outputs the starting points of the traversal.
+   * @return the ops needed to calculate {@code outputs}, not including {@code outputs}
+   */
+  public synchronized Set<GraphOperation> upstreamOps(Set<GraphOperation> outputs) {
```
Comment (Craigacp) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:346]: I missed this earlier, but why are these methods `synchronized`? It seems like they don't mutate the state of the Graph.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java [hunk 11]
```diff
@@ -304,6 +336,80 @@ private GraphOperation graphOp(Operand<?> operand) {
...
+
+  /**
+   * Get all ops directly or indirectly required to calculate {@code outputs} (not including {@code outputs}), including
+   * control dependencies.
+   *
+   * @param outputs the starting points of the traversal.
+   * @return the ops needed to calculate {@code outputs}, not including {@code outputs}
+   */
+  public synchronized Set<GraphOperation> upstream(Set<Operand<?>> outputs) {
+    return upstreamOps(outputs.stream().map(this::graphOp).collect(Collectors.toSet()));
```
Comment (Craigacp) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:391]: What does this do if you hand it a set of `Operand`s from `EagerOperation`s?

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java (hunk 3 above)
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java]: Ok. I'm still not sure that the semantics of this is quite right, but given we've not stabilised the API yet lets add it in and use it a bit first before making further decisions.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java (hunk 10 above)
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:346]: It doesn't, but if you mutate the graph while they are running the results could be incorrect.  More for downstream (with consumers) though, like if you added a new op using one of the inputs that had already been processed.  I can remove that from upstream.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java (hunk 9 above)
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:283]: They should be LinkedList too, I think, for performance.  All I'm doing is front removal and back append, no indexing.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java (hunk 11 above)
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:391]: `graphOp` will throw (with a nice message), since it calls `checkInput`.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java (hunk 10 above)
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java]: Ok.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java (hunk 9 above)
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:283]: The performance of `LinkedList` is bad due to having to do two pointer chases to find any given object and thus likely induce two cache misses (one for the node and one for the object itself). I'd expect `ArrayDeque` to be faster given it is based on a circular array, so it's likely to have the array always in cache when it's being traversed.

However performance isn't likely to be an issue in this code, so provided they are all the same I don't care.

Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java]: Polling takes about the same amount of dereferences, either `LinkedList -> Node -> Object` or `ArrayDeque -> backing array -> Object`.

**Review by Craigacp — DISMISSED** [review]
LGTM.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphOperation.java [hunk 12]
```diff
@@ -170,6 +184,155 @@ Tensor tensor(int outputIdx) {
     throw new IllegalStateException("Graph tensors must be fetched by running a session");
   }
 
+  /**
+   * Get the number of inputs to the op, not including control inputs.
+   */
+  public int numInputs() {
+    try (PointerScope scope = new PointerScope()) {
+      return TF_OperationNumInputs(getUnsafeNativeHandle());
+    }
```
Comment (karllessard) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphOperation.java:193]: We don't need a pointer scope here since we call straight a native endpoint without allocating anything

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java [hunk 13]
```diff
@@ -138,14 +219,205 @@ public GraphOperation operation(String name) {
...
+   * forbiddenNoInputBodyOps}) if null.
+   * @param forbiddenNoInputBodyOps types of ops to never allow as 0-input ops in the body.  Forbids all (except {@code
+   * allowedNoInputBodyOps}) if null.
+   * @return the set of operations needed to calculate outputs from inputs, including outputs and inputs
+   * @throws IllegalStateException if outputs depends on ops outside of the subgraph (i.e. is not calculable based
+   * solely on inputs)
+   * @see #completeSubgraph(Set, Set, boolean)
+   */
+  public synchronized Set<GraphOperation> completeSubgraph(Set<Operand<?>> inputs, Set<Operand<?>> outputs,
+      Set<String> allowedNoInputBodyOps, Set<String> forbiddenNoInputBodyOps) {
```
Comment (karllessard) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:265]: These two last extra parameters obfuscate a bit the overall solution, which would be very simple without them. Why can't we simply include all ops we encounter without this logic, whether they have inputs available or not, and let the caller filter them out after if he needs to?

Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:265]: Yeah, that's fine.  I don't like how easy it is to end up including things upstream from the inputs in the body then, but I don't see much of a way around it, since we can't just forbid it altogether.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java [hunk 14]
```diff
@@ -128,6 +138,77 @@ public GraphOperation operation(String name) {
...
+      if (operation == null) {
+        return null;
+      }
+      return new Output(operation, index);
+    } catch (NumberFormatException e) {
+      GraphOperation op = operation(output);
+      if (op == null) {
+        return null;
+      }
+      return new Output(op, 0);
```
[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java:189]: I'm not.  I think it was to handle cases where you had a `:` in the name but no output index, but that seems forbidden in the front end at least.  Tests pass though, so I think we're fine.

**Comment by karllessard** [comment]
@rnett looks good, are you done with the changes? If so, can you please check why the quick build has failed?

**Comment by rnett** [comment]
It's the javadoc issues in `framework`.  I think what's happening is `core-api` is set to ignore Javadoc failures, so Javadoc fails but the build succeeds, but then `framework`, which isn't set to ignore Javadoc failures, fails when it tries to depend on them.  I'm pretty sure this is an issue on master as well, I didn't add anything that should fail.

**Comment by karllessard** [comment]
Ok, I see. Indeed, the javadoc validation [is disabled](https://github.com/tensorflow/java/blob/0f0c638b691907e954e3f3bbff465a96c354a71d/tensorflow-core/tensorflow-core-api/pom.xml#L429) in the `tensorflow-core-api` since this generated documentation from Markdown is not 100% legit right now. 

Now what you are saying is that the framework documentation with links to the generated classes fail to pass this validation because it tries to validate the documentation it links to? 

I was not aware of that but I guess it's possible, @JimClarke5 can we clean up the framework documentation to at least fix the CI build, even if it means to disable these links in it? 

I will merge this PR then, thanks @rnett !

**Review by karllessard — APPROVED** [review]

**Comment by JimClarke5** [comment]
There is a C library for commonmark, but it would need to be extended with a JavaDoc renderer based on the HtmlRenderer. I have done this with Python and am now working on the Java version.  I could look at the C version and determine what needs to be done.

**Comment by karllessard** [comment]
Ok @JimClarke5 I'll let you judge how long it would take but for now the priority is really to fix the CI build so that snapshots get refreshed on Sonatype after merging a PR. Do you think we can simply fix the framework quickly by removing links to the generated ops?

**Comment by JimClarke5** [comment]
@karllessard  Let me experiment with the [CommonMark-cmark](https://github.com/commonmark/cmark) library. That would be the quickest fix and incorporate it into SourceWriter.cc

**Comment by rnett** [comment]
Couldn't we, for now, just let framework build w/ failing javadocs?  W/ a note to fix before the next release or something.

**Comment by JimClarke5** [comment]
@karllessard please explain exactly what you mean by “ quickly by removing links to the generated ops?”

**Comment by JimClarke5** [comment]
@karllessard As a test, I have been able to transform the markup from api_def_UnsortedSegmentMax.pbtxt

```
  description: <<END
Read
[the section on segmentation](https://tensorflow.org/api_docs/python/tf/math#Segmentation)
for an explanation of segments.

This operator is similar to the unsorted segment sum operator found
[(here)](../../../api_docs/python/math_ops.md#UnsortedSegmentSum).
Instead of computing the sum over segments, it computes the maximum such that:

\\(output_i = \max_{j...} data[j...]\\) where max is over tuples `j...` such
that `segment_ids[j...] == i`.

If the maximum is empty for a given segment ID `i`, it outputs the smallest
possible value for the specific numeric type,
`output[i] = numeric_limits<T>::lowest()`.

If the given segment ID `i` is negative, then the corresponding value is
dropped, and will not be included in the result.
```

to

```
/**
 * Read <a href="https://tensorflow.org/api_docs/python/tf/math#Segmentation">the section on segmentation</a> for an  
 * explanation of segments.This operator is similar to the unsorted segment sum operator found        
 * <a href="../../../api_docs/python/math_ops.md#UnsortedSegmentSum">(here)</a>.Instead of computing the sum over
 *  segments, it computes the maximum such that:\(output_i = \max_{j...} data[j...]\) where max is over tuples {@code j... } 
 * such that {@code segment_ids[j...] == i }.If the maximum is empty for a given segment ID {@code i }, it outputs the 
 * smallestpossible value for the specific numeric type,{@code output[i] = numeric_limits<T>::lowest() }.If the given segment 
 * ID {@code i } is negative, then the corresponding value isdropped, and will not be included in the result.
 * /

```
The only issue so far is the relative URL `../../../api_docs/python/math_ops.md#UnsortedSegmentSum` in the second link.
With a little brute force, I think I can change that to `{@link MathOps#unsupportedSegmentSum}`.

I can change SourceWriter.cc to call out to new C methods to handle the JavaDoc rendering and in turn link in the cmark C library. I will also have to add the external cmark source to` bazel_bin-external` ( I think ). Is external pulled down from GitHub somewhere? Also, there may be some edge cases I have to handle as I expand out to the full api_def directory.

**Comment by karllessard** [comment]
@JimClarke5 , I would say don't put too much effort of this C++ markdown converter because @rnett is working on porting the op generator to Java anyway (#244). Unless Ryan you think that we could use it to convert the documentation as we generate the new ops.pb file?

**Comment by rnett** [comment]
I mean, we probably could, but given this is a Java library originally (if I understood things right) it's probably easier to wait.
